### PR TITLE
4.x cacheengine

### PR DESCRIPTION
**Pull request by ADmad**

### CODE REVIEW

File: src/Cache/CacheEngine.php [hunk 1]
```diff
@@ -94,20 +98,25 @@ protected function ensureValidKey($key)
...
-     * @throws \Cake\Cache\InvalidArgumentException When the keys are not valid.
+     * @throws \Cake\Cache\InvalidArgumentException
      */
-    protected function ensureValidKeys($keys)
+    protected function ensureValidType($iterable, string $check = self::CHECK_VALUE)
     {
-        if (!is_array($keys) && !($keys instanceof \Traversable)) {
-            throw new InvalidArgumentException('A cache key set must be either an array or a Traversable.');
+        if (!is_iterable($iterable)) {
+            throw new InvalidArgumentException('A cache set / keys set must be either an array or a Traversable.');
```
Comment (markstory) [inline src/Cache/CacheEngine.php]: This error message is confusing to me. Can we use `$check` to make it easier for the user to know which mode failed?

File: tests/TestCase/Cache/Engine/ArrayEngineTest.php [hunk 2]
```diff
@@ -279,4 +281,62 @@ public function testAdd()
...
+     */
+    public function testWriteMany()
+    {
+        $data = new \ArrayObject([
+            'a' => 1,
+            'b' => 'foo',
+        ]);
+
+        $result = Cache::writeMany($data, 'array');
+        $this->assertTrue($result);
```
Comment (markstory) [inline tests/TestCase/Cache/Engine/ArrayEngineTest.php]: Should you also check that the object can be fetched out of the cache?

File: tests/TestCase/Cache/Engine/ArrayEngineTest.php [hunk 3]
```diff
@@ -279,4 +281,62 @@ public function testAdd()
         $result = Cache::add('test_add_key', 'test data 2', 'array');
         $this->assertFalse($result);
     }
+
+    /**
+     * Test writeMany() with Traversable
+     *
+     * @return void
+     */
+    public function testWriteMany()
```
Comment (markstory) [inline tests/TestCase/Cache/Engine/ArrayEngineTest.php]:
```suggestion
    public function testWriteManyTraversable()
```

**Comment by ADmad** [comment]
@markstory updates done as per feedback.

**Pull request merged**